spring/narayana-spring-boot/src/main/java/org/jboss/narayana/quickstart/spring/QuickstartService.java

```diff
@@ -0,0 +1,119 @@
+ * contributors by the @authors tag. See the copyright.txt in the
+ * distribution for a full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.narayana.quickstart.spring;
+
+import com.arjuna.ats.jbossatx.jta.RecoveryManagerService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.TransactionManager;
+import javax.transaction.xa.XAResource;
+
+/**
+ * Quickstart execution service.
+ *
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+@Service
+public class QuickstartService {
+
+    @Autowired
+    private TransactionManager transactionManager;
+
+    @Autowired
+    private EntriesService entriesService;
+
+    @Autowired
+    private MessagesService messagesService;
+
+    @Autowired
+    private RecoveryManagerService recoveryManagerService;
+
+    /**
+     * Write entry to the database, send JMS message, and commit the transaction.
+     *
+     * @param entry String entry to be saved to the database.
+     * @throws Exception
+     */
+    public void demonstrateCommit(String entry) throws Exception {
+        executeDemonstration(entry, transactionManager::commit, null);
+    }
+
+    /**
+     * Write entry to the database, send JMS message, and rollback the transaction.
+     *
+     * @param entry String entry to be saved to the database.
+     * @throws Exception
+     */
+    public void demonstrateRollback(String entry) throws Exception {
+        executeDemonstration(entry, transactionManager::rollback, null);
+    }
+
+    /**
+     * Write entry to the database, send JMS message, and crash application.
+     *
+     * @param entry String entry to be saved to the database.
+     * @throws Exception
+     */
+    public void demonstrateCrash(String entry) throws Exception {
+        executeDemonstration(entry, transactionManager::commit, new DummyXAResource(true));
+    }
+
+    /**
+     * Execute a recovery to complete transactions interrupted by the previous crash.
+     *
+     * @throws Exception
+     */
+    public void demonstrateRecovery() throws Exception {
+        System.out.println("Entries at the start: " + entriesService.getAll());
+        recoveryManagerService.addXAResourceRecovery(new DummyXAResourceRecovery());
+        Thread.sleep(5000);
+        System.out.println("Entries at the end: " + entriesService.getAll());
```

Comment:
Also a good point. Although its configured to be max of 3 seconds (+ execution of modules time) it would be worth adding either a guard or a note.

I suppose as we know the module will be called you could add a sync on the DummyXAResourceRecovery and do a .wait()/notify()